dash/include/dash/coarray/Utils.h

```diff
@@ -0,0 +1,212 @@
+      if(el != root){
+        dart_recv(&buffer, 1, dart_datatype<char>::value,
+                  tag,
+                  global_unit_t{el}
+        );
+      }
+    });
+  } else {
+    // I am a leave, send message to root
+    dart_send(&buffer, 1, DART_TYPE_BYTE,
+              tag,
+              global_unit_t{root});      
+  }
+
+  // Second phase: recieve message from root
+  DASH_LOG_DEBUG("Begin second phase of sync_images");
+  if(myid == root){
+    std::for_each(image_ids.begin(), image_ids.end(), [&](const element & el)
+    {
+      if(el != root){
+        dart_send(&buffer, 1, dart_datatype<char>::value,
+                  tag,
+                  global_unit_t{el}
+        );
+      }
+    });
+  } else {
+    dart_recv(&buffer, 1, dart_datatype<char>::value,
+              tag,
+              global_unit_t{root}
+    );
+  }
+}
+
+/**
+ * Broadcasts the value on master to all other members of this co_array
+ * \note fortran defines this function only for scalar Coarray.
+ *       This implementation allows you to broadcast arrays as well
+ * 
+ * \param coarr  coarray which should be broadcasted
+ * \param master the value of this unit will be broadcastet
+ * 
+ * \ingroup DashCoarrayLib
+ */
+template<typename T>
+void cobroadcast(Coarray<T> & coarr, const team_unit_t & master){
+  using value_type        = typename Coarray<T>::value_type;
+  const dash::dart_storage<value_type> ds(coarr.local_size());
+  DASH_ASSERT_RETURNS(
+    dart_bcast(coarr.lbegin(),
+               ds.nelem,
+               ds.dtype,
+               master,
+               coarr.team().dart_id()),
+    DART_OK);
+}
+
+/**
+ * Performes a broadside reduction of the Coarray images.
+ * \param coarr   perform the reduction on this array
+ * \param op      reduce operation
+ * \param master  unit which recieves the result. -1 to broadcast to all units
+ * 
+ * \ingroup DashCoarrayLib
+ */
+template<typename T, typename BinaryOp>
+void coreduce(Coarray<T> & coarr,
+              const BinaryOp &op,
+              team_unit_t master = team_unit_t{-1})
+{
+  using value_type = typename Coarray<T>::value_type;
+  using index_type = typename Coarray<T>::index_type;
+
+  constexpr auto ndim     = Coarray<T>::ndim();
+  const auto team_dart_id = coarr.team().dart_id();
+  bool broadcast_result   = (master < 0);
+  if(master < 0){master = team_unit_t{0};}
+
+  // position of first element on master
+  const auto & global_coords = coarr.pattern().global(master,
+                                 std::array<index_type,ndim> {});
+  const auto & global_idx    = coarr.pattern().at(global_coords);
+
+  const auto dart_gptr       = (coarr.begin() + global_idx).dart_gptr();
+  const dash::dart_storage<value_type> ds(coarr.local_size());
+
+  // as source and destination location is equal, master must not contribute
+  // in accumulation as otherwise it's value is counted twice
+  if(coarr.team().myid() != master){
+    DASH_ASSERT_RETURNS(
+      dart_accumulate(
```

Comment:
1) Note that  `dart_accumulate` performs an element-wise atomic update so it is unsafe for anything other than basic types. I suggest limiting `coreduce` to basic types for now.
2) Also, op is limited to predefined operations, which should at least be documented.
3) It should be explicitly documented that the result will be the first value on the receiving unit(s).

Reply:
1 and 2 are fixed now. What do you mean with 3? The function performs a broadside reduction:

```
Unit 0 | 1 3 0 2
Unit 1 | 0 2 1 1
----------------
All    | 1 5 1 3 (using coreduce(x,+))
```

Reply:
You're right, I wrote 3) together with a bunch of other comments that I later deleted when I realized that it's an element-wise reduction, that one should have been gone as well.